From f6147863750f45ebc3d29d3e61334938c0e8be59 Mon Sep 17 00:00:00 2001
From: "Ma, Liangliang" <liangliang.ma@intel.com>
Date: Sat, 8 Oct 2022 01:26:15 +0000
Subject: [PATCH 1/2] cmyk format

---
 jdcolor.c | 31 +++++++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/jdcolor.c b/jdcolor.c
index e5c7b58ebfad..d3281b61983a 100644
--- a/jdcolor.c
+++ b/jdcolor.c
@@ -532,6 +532,35 @@ rgb_rgb_convert(j_decompress_ptr cinfo, _JSAMPIMAGE input_buf,
   }
 }
 
+/*
+ * Convert CMYK to RGBA
+ */
+METHODDEF(void)
+cmyk_rgba_convert(j_decompress_ptr cinfo, _JSAMPIMAGE input_buf,
+    JDIMENSION input_row, _JSAMPARRAY output_buf, int num_rows)
+{
+    register _JSAMPROW inptr0, inptr1, inptr2, inptr3;
+    register _JSAMPROW outptr;
+    register JDIMENSION col;
+    JDIMENSION num_cols = cinfo->output_width;
+
+    while (--num_rows >= 0) {
+        inptr0 = input_buf[0][input_row];
+        inptr1 = input_buf[1][input_row];
+        inptr2 = input_buf[2][input_row];
+        inptr3 = input_buf[3][input_row];
+        input_row++;
+        outptr = *output_buf++;
+        for (col = 0; col < num_cols; col++) {
+            int k_diff = inptr3[col];
+            outptr[0] = (inptr0[col] * k_diff >> 8) + 1;
+            outptr[1] = (inptr1[col] * k_diff >> 8) + 1;
+            outptr[2] = (inptr2[col] * k_diff >> 8) + 1;
+            outptr[3] = 0xFF;
+            outptr += 4;
+        }
+    }
+}
 
 /*
  * Adobe-style YCCK->CMYK conversion.
@@ -853,6 +882,8 @@ _jinit_color_deconverter(j_decompress_ptr cinfo)
         cconvert->pub._color_convert = null_convert;
       else
         cconvert->pub._color_convert = rgb_rgb_convert;
+    } else if (cinfo->jpeg_color_space == JCS_CMYK) {
+        cconvert->pub._color_convert = cmyk_rgba_convert;
     } else
       ERREXIT(cinfo, JERR_CONVERSION_NOTIMPL);
     break;
-- 
2.38.1

